File: typescript/tssearch/tests/searchoptions_test.js
/// <reference path="../typings/node/node.d.ts"/>
/// <reference path="../src/searchoptions.ts"/>
/// <reference path="../src/searchsettings.ts"/>
/*
 * searchoptions_test.js
 *
 * Some nodeunit tests of searchoptions.js
 */
var SearchOptions = require('../build/searchoptions.js').SearchOptions;
var SearchSettings = require('../build/searchsettings.js').SearchSettings;
exports.testNoArgs = function (test) {
    var searchOptions = new SearchOptions();
    searchOptions.settingsFromArgs([], function (err, settings) {
        if (err) {
            test.ok(false, "There was an error calling settingsFromArgs: " + err);
            test.done();
        }
        test.ok(!settings.archivesOnly, "archivesOnly is false by default");
        test.ok(!settings.debug, "debug is false by default");
        test.ok(settings.excludeHidden, "excludeHidden is true by default");
        test.ok(!settings.firstMatch, "firstmatch is false by default");
        test.ok(settings.linesAfter === 0, "linesAfter == 0 by default");
        test.ok(settings.linesBefore === 0, "linesBefore == 0 by default");
        test.ok(!settings.listDirs, "listDirs is false by default");
        test.ok(!settings.listFiles, "listFiles is false by default");
        test.ok(!settings.listLines, "listLines is false by default");
        test.ok(settings.maxLineLength == 150, "maxLineLength == 150 by default");
        test.ok(!settings.multilineSearch, "multilineSearch is false by default");
        test.ok(settings.printResults, "printResults is set to true in settingsFromArgs");
        test.ok(!settings.printUsage, "printUsage is false by default");
        test.ok(!settings.printVersion, "printVersion is false by default");
        test.ok(settings.recursive, "recursive is true by default");
        test.ok(!settings.searchArchives, "searchArchives is false by default");
        test.ok(settings.startPath === '', "startPath is empty by default");
        test.ok(!settings.uniqueLines, "uniqueLines is false by default");
        test.ok(!settings.verbose, "verbose is false by default");
        test.done();
    });
};
exports.testValidArgs = function (test) {
    var searchOptions = new SearchOptions();
    var args = ['-x', 'js,java', '-s', 'Searcher', '.'];
    searchOptions.settingsFromArgs(args, function (err, settings) {
        if (err) {
            test.ok(false, "There was an error calling settingsFromArgs: " + err);
            test.done();
            return;
        }
        test.ok(settings.inExtensions.length === 2, "inExtensions has two extensions");
        test.ok(settings.inExtensions[0] === 'js', "first inExtension is js");
        test.ok(settings.inExtensions[1] === 'java', "first inExtension is java");
        test.ok(settings.searchPatterns.length === 1, "searchPatterns has one pattern");
        test.ok(settings.searchPatterns[0].source === 'Searcher', "pattern === Searcher");
        test.done();
    });
};
exports.testArchivesOnly = function (test) {
    var searchOptions = new SearchOptions();
    var args = ['--archivesonly'];
    searchOptions.settingsFromArgs(args, function (err, settings) {
        if (err) {
            test.ok(false, "There was an error calling settingsFromArgs: " + err);
            test.done();
        }
        test.ok(settings.archivesOnly === true, "archivesOnly is true");
        test.ok(settings.searchArchives === true, "searchArchives is true");
        test.done();
    });
};
exports.testDebug = function (test) {
    var searchOptions = new SearchOptions();
    var args = ['--debug'];
    searchOptions.settingsFromArgs(args, function (err, settings) {
        if (err) {
            test.ok(false, "There was an error calling settingsFromArgs: " + err);
            test.done();
        }
        test.ok(settings.debug === true, "debug is true");
        test.ok(settings.verbose === true, "verbose is true");
        test.done();
    });
};
exports.testMissingArg = function (test) {
    var searchOptions = new SearchOptions();
    var args = ['-x'];
    searchOptions.settingsFromArgs(args, function (err, settings) {
        if (err) {
            var expected = "Missing argument for option x";
            test.ok(err.message === expected, "Got missing argument err");
            test.done();
        }
        else {
            test.ok(false, "Did not get expected missing argument error");
            test.done();
        }
    });
};
exports.testInvalidArg = function (test) {
    var searchOptions = new SearchOptions();
    var args = ['-Q'];
    searchOptions.settingsFromArgs(args, function (err, settings) {
        if (err) {
            var expected = "Invalid option: Q";
            test.ok(err.message === expected, "Got invalid option err");
            test.done();
        }
        else {
            test.ok(false, "Did not get expected invalid option error");
            test.done();
        }
    });
};
exports.testSettingsFromJson = function (test) {
    var searchOptions = new SearchOptions();
    var settings = new SearchSettings();
    var json = '{\n' +
        '  "startpath": "~/src/xsearch/",\n' +
        '  "in-ext": ["js","ts"],\n' +
        '  "out-dirpattern": "node_module",\n' +
        '  "out-filepattern": ["temp"],\n' +
        '  "searchpattern": "Searcher",\n' +
        '  "linesbefore": 2,\n' +
        '  "linesafter": 2,\n' +
        '  "debug": true,\n' +
        '  "allmatches": false,\n' +
        '  "includehidden": true\n' +
        '}';
    var err = searchOptions.settingsFromJson(json, settings);
    test.ok(err == null, "Null err");
    test.ok(settings.startPath === '~/src/xsearch/', "Startpath === ~/src/xsearch/");
    test.ok(settings.inExtensions.length === 2, "settings.inExtensions.length === 2");
    test.ok(settings.outDirPatterns.length === 1, "settings.outDirPatterns.length === 1");
    test.ok(settings.outDirPatterns[0].source === 'node_module', "settings.outDirPatterns[0].source === 'node_module'");
    test.ok(settings.outFilePatterns.length === 1, "settings.outFilePatterns.length === 1");
    test.ok(settings.outFilePatterns[0].source === 'temp', "settings.outFilePatterns[0].source === 'temp'");
    test.ok(settings.searchPatterns.length === 1, "settings.searchPatterns.length === 1");
    test.ok(settings.searchPatterns[0].source === 'Searcher', "settings.searchPatterns[0].source === 'Searcher'");
    test.ok(settings.linesBefore === 2, "settings.linesBefore === 2");
    test.ok(settings.linesAfter === 2, "settings.linesAfter === 2");
    test.ok(settings.debug, "settings.debug === true");
    test.ok(settings.verbose, "settings.verbose === true");
    test.ok(settings.firstMatch, "settings.firstMatch === true");
    test.ok(settings.excludeHidden === false, "settings.excludeHidden === false");
    test.done();
};

File: go/gosearch/pkg/gosearch/searchoptions_test.go
package gosearch

import "testing"

func TestSearchSettingsFromNoArgs(t *testing.T) {
	searchOptions := NewSearchOptions()

	var args []string

	settings, err := searchOptions.SearchSettingsFromArgs(args)
	if err != nil {
		t.Errorf("SearchSettingsFromArgs: err: %v", err)
	}

	if settings.ArchivesOnly() ||
		settings.Debug() ||
		settings.FirstMatch() ||
		settings.IncludeHidden() ||
		settings.MultiLineSearch() ||
		settings.PrintDirs() ||
		settings.PrintFiles() ||
		settings.PrintLines() ||
		!settings.PrintResults() ||
		settings.PrintUsage() ||
		settings.PrintVersion() ||
		!settings.Recursive() ||
		settings.SearchArchives() ||
		settings.UniqueLines() ||
		settings.Verbose() {
		t.Errorf("settings did not match defaults")
	}
}

func TestSearchSettingsFromValidArgs(t *testing.T) {
	searchOptions := NewSearchOptions()

	args := []string{
		"-x", "go", "-s", "Searcher", ".",
	}

	settings, err := searchOptions.SearchSettingsFromArgs(args)
	if err != nil {
		t.Errorf("SearchSettingsFromArgs: err: %v", err)
	}

	if len(settings.Paths()) != 1 {
		t.Errorf("settings.Paths is empty")
	}

	if settings.Paths()[0] != "." {
		t.Errorf("settings.Paths[0] != \".\"")
	}

	if len(settings.InExtensions()) != 1 {
		t.Errorf("len(settings.InExtensions) = %d, expected 1",
			len(settings.InExtensions()))
	}
	expectedExt := "go"
	if settings.InExtensions()[0] != expectedExt {
		t.Errorf("settings.InExtensions[0] (\"%s\") != \"%s\"",
			settings.InExtensions()[0], expectedExt)
	}
}

func TestSearchSettingsFromJson(t *testing.T) {
	searchOptions := NewSearchOptions()

	jsonSettings := []byte(`{
  "path": "~/src/xsearch/",
  "in-ext": ["js","ts"],
  "out-dirpattern": "node_module",
  "out-filepattern": ["temp"],
  "searchpattern": "Searcher",
  "linesbefore": 2,
  "linesafter": 2,
  "debug": true,
  "allmatches": false,
  "includehidden": true
}`)

	settings := GetDefaultSearchSettings()
	var err error
	err = searchOptions.SettingsFromJson(jsonSettings, settings)
	if err != nil {
		t.Errorf("TestSearchSettingsFromJson: err: %v", err)
	}

	if len(settings.Paths()) != 1 {
		t.Errorf("settings.Paths is empty")
	}

	if settings.Paths()[0] != "~/src/xsearch/" {
		t.Errorf("settings.Paths[0] != \"~/src/xsearch/\"")
	}

	if len(settings.InExtensions()) != 2 {
		t.Errorf("len(settings.InExtensions) = %d, expected 2",
			len(settings.InExtensions()))
	}
	expectedInExts := []string{"js", "ts"}
	for i, _ := range expectedInExts {
		if settings.InExtensions()[i] != expectedInExts[i] {
			t.Errorf("settings.InExtensions[%d] (\"%s\") != \"%s\"",
				i, settings.InExtensions()[i], expectedInExts[i])
		}
	}

	if settings.OutDirPatterns().Len() != 1 {
		t.Errorf("settings.OutDirPatterns().Len() = %d, expected 1",
			settings.OutDirPatterns().Len())
	}
	if settings.OutDirPatterns().Get(0).String() != "node_module" {
		t.Errorf("settings.OutDirPatterns().Get(0).String() (\"%s\") != \"node_module\"",
			settings.OutDirPatterns().Get(0).String())
	}

	if settings.OutFilePatterns().Len() != 1 {
		t.Errorf("settings.OutFilePatterns().Len() = %d, expected 1",
			settings.OutFilePatterns().Len())
	}
	if settings.OutFilePatterns().Get(0).String() != "temp" {
		t.Errorf("settings.OutFilePatterns.Get(0).String() (\"%s\") != \"temp\"",
			settings.OutFilePatterns().Get(0).String())
	}

	if settings.SearchPatterns().Len() != 1 {
		t.Errorf("len(settings.SearchPatterns.patterns) = %d, expected 1",
			settings.SearchPatterns().Len())
	}
	if settings.SearchPatterns().Get(0).String() != "Searcher" {
		t.Errorf("settings.SearchPatterns().Get(0).String() (\"%s\") != \"Searcher\"",
			settings.SearchPatterns().Get(0).String())
	}

	if settings.LinesBefore() != 2 {
		t.Errorf("settings.LinesBefore (%d) != 2", settings.LinesBefore)
	}

	if settings.LinesAfter() != 2 {
		t.Errorf("settings.LinesAfter (%d) != 2", settings.LinesAfter)
	}

	if !settings.Debug() {
		t.Errorf("settings.Debug (%t) != true", settings.Debug)
	}

	if !settings.FirstMatch() {
		t.Errorf("settings.FirstMatch (%t) != true", settings.FirstMatch)
	}

	if !settings.IncludeHidden() {
		t.Errorf("settings.IncludeHidden (%t) != true", settings.IncludeHidden)
	}
}
